purge verification experiments of ECCO_CTRL_DEPRECATED code

mjlosch opened the pull request:
## What changes does this PR introduce?
As first step of an effort to remove ECCO_CTRL_DEPRECATED code, this PR replaces ECCO_CPPOPTIONS.h and ECCO_CTRL_DEPRECATED code in the remaining verification experiments.

## What is the current behaviour? 
many verification experiments still use ECCO_CPPOPTIONS.h and/or have ECCO_CTRL_DEPRECATED defined.

## What is the new behaviour 
These remaining experiments now use AUTODIFF/CTRL/COST/ECCO_OPTIONS.h with generic control variables and generic ecco cost functions (where applicable) instead of ECCO_CPPOPTIONS.h and have ECCO_CTRL_DEPRECATED code undefined:

- [ ] 1D_ocean_ice_column
- [ ] OpenAD
- [ ] bottom_ctrl_5x5, waiting for #470
- [ ] global_ocean.cs32x15 (still compiles and runs both DEPRECATED code and generic control variables)
- [ ] global_with_exf
- [ ] isomip
- [ ] lab_sea
- [ ] obcs_ctrl
- [ ] offline_exf_seaice
- [ ] tutorial_dic_adjoffline
- [ ] tutorial_global_oce_biogeo
- [ ] tutorial_global_oce_optim

These OpenAD experiments still need to be switched, maybe in a separate PR
- OpenAD
- global_ocean.90x40x15
- isomip
- tutorial_global_oce_biogeo
- tutorial_global_oce_optim
- tutorial_tracer_adjsens

## Does this PR introduce a breaking change? 
Not intentionally, but it will mean that ECCO_CTRL_DEPRECATED code will no longer be tested and not maintained. Soon after the PR, we may get rid of that code altogether.

## Other information:
There are still a few issues where I need help, which is why this is a draft:
1. In some cases I had to define ALLOW_UNPACK_METHOD2 to make this work: 1D_ocean_ice_column, lab_sea, tutorial_global_oce_optim (two others had this flag defined already: global_oce_biogeo_bling, global_ocean.cs32x15), because there is something fishy with the default methods and reading the weights file, probably because we use MDSREADFIELD_2D_GL. Need to be sorted out (or at least understood).
2. in some cases it is not clear to me if the generic cost function does the same as the non-generic ones (e.g. weights for control variable constraints)
3. for tutorial_dic_adjoffline there are unclear additional RECOMPUTATION warnings that either need to be ignored or dealt with (to be discussed)
4. I am waiting for #470 to be merged, because there, part of the switch of bottom_ctrl_5x5 has already been done.
5. for lab_sea, I was not yet able to get the correct MDT cost function contribution to work; the contributions of theta, salt, and sst are the same as before.

## Suggested addition to `tag-index`
o modify remaining verification experiments so that they no longer use ECCO_CPPOPTIONS.h and have ECCO_CTRL_DEPRECATED defined

timothyas commented on pkg/grdchk/grdchk_getadxx.F:
```diff
@@ -305,15 +305,7 @@ subroutine grdchk_getadxx(
          write(fname(1:80),'(80a)') ' '
          write(fname(1:80),'(3a,i10.10)')
      &        yadmark, xx_bottomdrag_file(1:il),'.',optimcycle
-#endif /* ALLOW_HFLUXM_CONTROL */
-
-#ifdef ALLOW_HFLUXM_CONTROL
-      else if ( grdchkvarindex .eq. 24 ) then
-         il=ilnblnk( xx_hfluxm_file )
-         write(fname(1:80),'(80a)') ' '
-         write(fname(1:80),'(3a,i10.10)')
-     &        yadmark, xx_hfluxm_file(1:il),'.',optimcycle
-#endif /* ALLOW_HFLUXM_CONTROL */
```
It's not clear to me why this block (and similar in the getxx, setxx) should be deleted. Is this why the `tutorial_global_oce_optim` is not passing? That experiment uses xx_hfluxm, so it seems like a candidate...

mjlosch replied:
Yes, you are probably right about that. I removed all HFLUXM code, as it is specific to `tutorial_global_oce_optim`, and this experiment works for TAF, but I haven't yet converted any of the OpenAD experiments (which I will, once all TAF experiments work). The switch for the OpenAD experiments should be straightforward, shouldn't it?

mjlosch replied:
Straying from my original plan, I tried to change `tutorial_global_oce_optim/code_oad`, which didn't work, so I'll revert that change.

mjlosch replied:
reverted to please OpenAD

timothyas commented on verification/lab_sea/code_ad/CPP_OPTIONS.h:
```diff
@@ -133,7 +133,7 @@ C   Although this method, inherited from ECCO setup, has been traditionally
 C   used for all adjoint built, work is in progress to allow to use the
 C   standard method also for adjoint built.
 #ifdef PACKAGES_CONFIG_H
-# include "ECCO_CPPOPTIONS.h"
+c# include "ECCO_CPPOPTIONS.h"
 #endif
```
Super minor, but in my opinion this could be deleted (and the comments above it) rather than carrying around commented code and macros.

mjlosch replied:
You are right. Once there is no more ECCO_CPPOPTIONS.h, we should probably remove this line/piece of code from all CPP_OPTIONS.h files.

jm-c replied:
I am fine too with leaving this include commented out for now (as it was in model/inc version) and removing all these commented out include later on (the advantage would be to reduce number of CPP_OPTIONS.h affected by this PR).

timothyas commented on verification/lab_sea/input_ad/data.ctrl:
```diff
@@ -5,102 +5,102 @@
- xx_sssstartdate2 = 000000,
- xx_sssperiod     = 864000.0,
- xx_sss_file      = 'xx_sss',
+# #
+#  xx_theta_file      ='xx_theta',
+#  xx_salt_file       ='xx_salt',
+#  xx_siarea_file     ='xx_siarea',
+#  xx_siheff_file     ='xx_siheff',
```
Also very minor, this all could get deleted rather than commented.

mjlosch replied:
agreed, specifically this experiment is not yet complete, as I haven't manage to get the same cost function for the generic cost function code, yet (see initial statement in "Other information"). Once this is done, I'll clean up.

mjlosch replied:
Done.

timothyas commented on verification/tutorial_dic_adjoffline/code_ad/CPP_OPTIONS.h:
```diff
@@ -119,7 +119,7 @@ C   Although this method, inherited from ECCO setup, has been traditionally
 C   used for all adjoint built, work is in progress to allow to use the
 C   standard method also for adjoint built.
 #ifdef PACKAGES_CONFIG_H
-# include "ECCO_CPPOPTIONS.h"
+c# include "ECCO_CPPOPTIONS.h"
 #endif
```
Candidate for removal ... :)

mjlosch commented:
@antnguyen13 I could use some help with lab_sea. Do you think that it is possible to reproduce the results of the lab_sea AD experiments just using the generic cost functions of the ecco pkg? What I have tried is insufficient (it gives me something like `0.36D+03` instead of `objf_hmean = 0.5026D-01`, but the code leading to `objf_hmean` is so convoluted that I don't know what's going on there). We don't need to use cost_gencost_sshv4.F for this, I hope, as this also contains a lot of hardwired stuff ... a bit lost ...

jm-c commented on pkg/ecco/cost_generic.F:
```diff
@@ -396,21 +406,40 @@ subroutine cost_genloop(
         if ( domean.OR.doanom )
-     &    call ecco_addmask(localdif,difmask, nnzobs,localdifsum,
-     &    localdifnum, nnzobs,myThid)
+     &    call ecco_addmask(l ocaldif, difmask,nnzobs, localdifsum,
```
this splitted "l ocaldif" looks suspicious to me

mjlosch replied:
yeah, that's not right, but gfortran can deal with this. I'll fix it.

mjlosch commented:
Since 721cb6a all arrays in ecco_toolbox.F, cost_genread.F are now declared with Nr for the third dimension. This is necessary to make lab_sea work properly with the generic cost function machinery (I think), but there are many (untested) places, where these routines are called with 2D fields as arguments, e.g. in cost_gencost_ssh4.F, cost_gencost_seaicev4.F, cost_gencost_bpv4.F, cost_gencost_transp.F. Cleaning this up will not be straightforward, I am afraid, so I would be extremely grateful if this could be tested in an eccov4 setup (maybe @owang01 could try this?). 

I think that if I pass a 2D array `array2d(:,:,nSx,nSy)` to a routine that expects a 3D array `array3d(:,:,Nr,nSx,nSy)`, but only use `k=1` (as specified in the argument list), it should still work, but I cannot be sure. It's definitely not clean, but if it works, I would like to leave it as it is. I think it was definitely wrong before (passing 3D arrays with parameter `Nr` to routines with dynamic array allocation/declaration, as it was done in cost_generic.F before 721cb6a).

jm-c commented:
@mjlosh regarding your question, we will need to check each of these call to any of the S/R you changed (17 in toolbox + cost_genread) to make sure the code is right.

jm-c commented:
and just as an example, calls to ECCO_READWEI in cost_gencost_sshv4.F are not right now (I think). Some of these S/R might be used in AD experiments cs32 & llc90 from verification_other, so you might be able to check.
But again, will also need to make sure the un-tested ones are not broken.

mjlosch commented:
> and just as an example, calls to ECCO_READWEI in cost_gencost_sshv4.F are not right now (I think).

@jm-c I see what you mean. How would one fix this? All I can think of involves extra subroutines for 2d or 3d.

mjlosch commented:
@owang01 the recent commit f47dd59 should fix the major issues that I anticipate with 2D routines like `cost_gencost_sshv4.F`
to be continued ...

owang01 commented:
@mjlosch There are missing arguments when calling ecco_readwei. 
cost_gencost_sshv4.F: lines 220-221 should be 
     &     call ecco_readwei(gencost_errfile(igen_mdt),
     &     gencost_weight(:,:,1,1,igen_mdt),**1,1,**1,dosumsq,myThid)
cost_gencost_bpv4.F: lines 99-100 should be 
     &   call ecco_readwei(gencost_errfile(kgen),
     &     gencost_weight(:,:,1,1,kgen),**1,1,**1,dosumsq,myThid)

mjlosch commented:
> @mjlosch There are missing arguments when calling ecco_readwei.
> cost_gencost_sshv4.F: lines 220-221 should be
> & call ecco_readwei(gencost_errfile(igen_mdt),
> & gencost_weight(:,:,1,1,igen_mdt),1,1,1,dosumsq,myThid)
> cost_gencost_bpv4.F: lines 99-100 should be
> & call ecco_readwei(gencost_errfile(kgen),
> & gencost_weight(:,:,1,1,kgen),1,1,1,dosumsq,myThid)

Hi @owang01 thanks, I just fixed this. The current code works for verification_other/global_oce_cs32. I will also try verification_other/global_oce_llc90, which usually takes a little longer to run, but it would be great, if you could check (again?) with a current "ecco production" like configuration.

mjlosch commented:
On ollie0.awi.de and ifort, I am getting this for verification_other/global_oce_llc90:
```
Y Y Y Y 16>16< 5 16 14 14 14 16 16 16 14 13 11 11 12 13 13 11 12 pass  global_oce_llc90  (e=0, w=15, lfd=1, dop=1, sm=1)
Y Y Y Y 16>16< 1 16 16 16 14 16 16 14 13 13 12 11 12 13 13 10 12 pass  global_oce_llc90.core2
```
(and then my job ran out of time, and now there's a TAF problem, so I can't restart it). I don't know what to expect for the FD values, but otherwise this looks good to me.

owang01 commented:
@mjlosch I checked it with the ECCO V4r5 configuration for a 35-day forward simulation. The fractional cost difference is ~1e-4. I wonder consider the commit is fine. 

I have two more comments:
1) In ECCO_CPRSRL, should "Nr" be replaced by "nzIn" in the declaration of fldIn and fldOut as it was done in ECCO_CP
 https://github.com/MITgcm/MITgcm/blob/b79be64d3509d3c639fbea62a740fd6355c91269/pkg/ecco/ecco_toolbox.F#L380-L381
2) Should ECCO_MULTFIELD be updated as well (although it is not being used)? If so, the arguments would be rearraneged to fld, fld2, nnzloc, myThid 
https://github.com/MITgcm/MITgcm/blob/b79be64d3509d3c639fbea62a740fd6355c91269/pkg/ecco/ecco_toolbox.F#L815
Similar to 1), "Nr" in the declaration of fld, fld2 should be replaced by "nnzloc"
https://github.com/MITgcm/MITgcm/blob/b79be64d3509d3c639fbea62a740fd6355c91269/pkg/ecco/ecco_toolbox.F#L831-L832

mjlosch commented:
@owang01 thanks for checking and finding problems. I'll fix that on Monday. Does it make any difference for your simulations if you (correctly) replace `Nr` by `nzIn`? I'd be surprised, if not.

About the unused functions: I have left them untouched but just added a stop, so that anyone who wants to use them will be forced to do something (and fix the routine). If we never use a routine, we might as well abandon it, but I can be convinced otherwise. I'd like to hear other opinions about that (e.g. @jm-c @ifenty @gaelforget @antnguyen13)

To be honest, I would expect zero difference for the cost function (I can achieve that with lab_sea), so a non-zero difference may point a problem somewhere (and I am pretty sure the code is not correct in master, and only works if (1) you don't use 2D field with cost_generic and (2) you don't use multitheading, i.e. bi=bj=1 always). But if you find a relative difference of 1e-4 (0.1 %) OK, I am fine with that. I just don't want to jeopardise any ECCO production runs.

mjlosch commented:
@owang01 I have modified ecco_toolbox.F some more following your suggestions. I also modified cost_gencost_seaicev4.F Could you please have a look and see if you can approve of that? (verification_other/global_oce_llc passes the tests)

owang01 commented:
@mjlosch Thanks for the changes. It looks good to me.

owang01 commented:
@mjlosch I want to clarify the relative cost difference I mentioned early. It is not caused by the changes made in this PR (#505). I was actually able to get identical costs using this PR and one recent commit before this PR https://github.com/MITgcm/MITgcm/commit/f11a35cbcd1678f420524492f31581d3572c8c78.

jm-c reviewed (commented):
There are many files that changed, and will take a look at some of them later on. 

For now, I can update the lab_sea output_adm*.txt (3) and the output_tlm.txt there (new output from villon) unless we decide to keep this experiment unchanged for now and update it in an other PR ?

Also, I think it would be good to mention (in tag-index addition) that experiment tutorial_global_oce_biogeo has now less TAF recomputation warnings (6 -> 2). And I don't know why the cost-fct valiue reported by testreport in experiment tutorial_dic_adjoffline does not change in ADM test (16 matching digits) but does in TLM test:
this PR: Y Y Y Y 12>14< 8 pass  tutorial_dic_adjoffline
master: Y Y Y Y 16>16< 8 pass  tutorial_dic_adjoffline

jm-c commented:
I will fix the conflicts in these 2 files (from merging in PR #508).

jm-c commented:
Ok, lab_sea  looks much better now with only small differences in forward gradient (something we should expect because of machine truncation).

jm-c commented on pkg/ecco/ecco_toolbox.F:
```diff
@@ -1086,74 +1045,100 @@ subroutine ecco_readbar(
-        else
-           call active_read_xyz( active_var_file, active_var,
+      ELSEIF ( nnzbar .EQ. Nr ) THEN
+       CALL ACTIVE_READ_XYZ( active_var_file, active_var,
```
This call is not right anytime ECCO_READBAR is called with nz <> nnzbar = Nr ; I suggest to change the condition to:
   ELSEIF ( nnzbar.EQ.Nr .AND. nz.EQ.Nr ) THEN

mjlosch replied:
Done

jm-c commented on pkg/ecco/ecco_toolbox.F:
```diff
@@ -1086,74 +1045,100 @@ subroutine ecco_readbar(
+       CALL READ_REC_XY_RL( active_var_file, tmpFld2D,
+     &                      iRec, 1, myThid )
+      ELSEIF ( nnzbar .EQ. Nr ) THEN
+       CALL READ_REC_XYZ_RL( active_var_file, active_var,
```
same issue here

mjlosch replied:
Done

jm-c commented:
@mjlosch I looked at "ecco_toolbox.F" changes, and have few minor suggestions:

1.  regarding un-used S/R ECCO_DIFFANOMMSK & ECCO_OBSMSK, the argument list has not been changed but some argument-variable size have changed. It's not very clear to me when any of these 2 S/R would be called so may be it would be simpler to not change them at all or remove them.
2. since other S/R argument list have changed, we might think of getting a more consistent argument order across S/R. For instance, ECCO_DIVFIELD (xx,yy) does update xx <-- xx/yy but ECCO_MULTFIELD(xx,yy) update yy <-- xx*yy ; personally, I would find it easier for these S/R to update the first argument (as ECCO_DIVFIELD does) , and would requires to change ECCO_ADD, ECCO_SUBTRACT and ECCO_MULTFIELD ; and could also move "multloc" and "numerloc" in second place in argument list of ECCO_MULT and ECCO_DIV to make them more similar to MULTFIELD & DIVFIELD ; but it does not need to be changed if you prefer not to.
3. regarding ECCO_READBAR, the error message written to "msgBuf" is not printed, so it's not very useful ; and renaming arguments "nz --> nzIn" and "nnzbar -> nLev" would make this S/R more similar to the other ones in this file.

Otherwise, in the code where these S/R are called:

1. would be cleaner to remove nnzsiv4 & nnzbar from cost_gencost_seaicev4.F and to remove nnzbar & nnzobs from cost_gencost_transp.F and use instead 1 and Nr respectively: these local parameters are not used consistently in these 2 S/R so it does not help to have a more complicated and inconsistent code.
2. in cost_generic.F, it seems that it should not be used with nnzobs > nnzbar ; should we add a warning there for this case ? Also, since there are 2 S/R in this file we could list both of them at the top of the file.

jm-c commented:
Unrelated to this PR, I noticed that:

1. in cost_gencost_transp.F, the number of tiles in X is not right (missing nSx) when setting "facNum"
2. in cost_gencost_seaicev4.F, lines 200-205 (in SEAICECOST_JPL) should be:
    gencost_weight(1-Olx,1-Oly,1,1.igen...) instead of gencost_weight(1,1,1,1,igen...)

But since this PR is changing these 2 files, tricky to make edits from master branch to fix this.

jm-c commented:
Regarding this:
> Unrelated to this PR, I noticed that:
> 
> 1. in cost_gencost_transp.F, the number of tiles in X is not right (missing nSx) when setting "facNum"
> 2. in cost_gencost_seaicev4.F, lines 200-205 (in SEAICECOST_JPL) should be:
>    gencost_weight(1-Olx,1-Oly,1,1.igen...) instead of gencost_weight(1,1,1,1,igen...)

WIll put this in a new PR once this one is merged.

mjlosch commented:
@jm-c 
> And I don't know why the cost-fct valiue reported by testreport in experiment tutorial_dic_adjoffline does not change in ADM test (16 matching digits) but does in TLM test:
> this PR: Y Y Y Y 12>14< 8 pass tutorial_dic_adjoffline
> master: Y Y Y Y 16>16< 8 pass tutorial_dic_adjoffline

It is actually funny that the test reports only so small deviations, because the tlm - gradients are very wrong, should be order 2e17, but they are 0.

mjlosch commented:
@jm-c 
For `tutorial_dic_adjoffline`, I find that the surface "pco2" computed in dic_surfforcing_tl is different between this branch and master, which then leads to a different "fluxco2" which then leads to a different cost function. I don't understand, why this does not show up in the trcstat_ptracerXX monitor output, because they should be forced by this flux, shouldn't they? In the master branch, the time dependent surface forcing fields/files have much fewer TLM versions than in this branch (eg. external_fields_load now has a tl version), which must have something to do with specifying GENARR2 instead of GEN2_CONTROL, but I am not sure if that matters.

I am a little lost, because I don't know the DIC code well enough to understand where this difference may come from.

mjlosch commented:
In tutorial_dic_adjoffline, the TL test cost function is now different from the AD cost and the ref TL (std) cost function value:
```
run (verif_ecco_ctrl_deprecated)> grep "ADM  ref" ../results/output_adm.txt | head -1
(PID.TID 0000.0001)  ADM  ref_cost_function      = -9.44034170278785E+11
run (verif_ecco_ctrl_deprecated)> grep "TLM  ref" output_tlm.std.txt | head -1
(PID.TID 0000.0001)  TLM  ref_cost_function      = -9.44034170278785E+11
run (verif_ecco_ctrl_deprecated)> grep "TLM  ref" output_tlm.txt | head -1 [this is the current (wrong) output]
(PID.TID 0000.0001)  TLM  ref_cost_function      = -9.44034170278446E+11
```
So this PR does introduce a bug.

jm-c commented:
@mjlosch I think you are right, and I can see small differences (of the order of cost fct differences) in the initial pH (at the end of dic_surfforcing_init_tl). Seems that the differences (at least for the initial pH) come from calling carbon_coeffs_tl instead of carbon_coeffs (as it is with deprecated code but also as it is with updated code when running the AD test).
At this point it's not clear why carbon_coeffs_tl and  carbon_coeffs give slightly different answer ; may be the amont of computations within the same do-loop bloc (double in case of _tl version) is treated differently by the compiler/machine and results in small differences ? then the potentially very sensitive pH computation could introduce these small differences ?
in cost-fct ?

jm-c commented:
@mjlosch I looked more carefully, and it turns out that TAF does change the FWD part of S/R CARBON_COEFFS when
evaluating akb(i,j,bi,bj) in carbon_coeffs_tl, changing the order in which terms are added (+ dropping 1 pair of parenthesis).
I manage to force TAF to keep the addition of terms in the same order by adding the following aux var:
            help_akb =
      &          (148.0248 _d 0 + 137.1942 _d 0*sqrts + 1.62142 _d 0*s)
and replace line 643 of carbon_gchem.F with simply:
      &         help_akb +
With this change, the addition of terms (for akb) is done in the same order in carbon_coeffs_tl and carbon_coeffs and the cost-function is unchanged (as well as pickup files).
So I think we know now where these differences come from and we could move towards merging this PR.
Do you agree ?
